src/Application/Items/Commands/BuyItemCommand.cs

```diff
@@ -59,11 +59,16 @@ public async Task<Result<UserItemViewModel>> Handle(BuyItemCommand req, Cancella
                 return new(CommonErrors.ItemNotBuyable(req.ItemId));
             }
 
-            if (user.Items.Any(ui => ui.ItemId == req.ItemId && ui.Rank == 0))
+            if (user.Items.Any(ui => ui.ItemId == req.ItemId && ui.Item!.Rank == 0))
             {
                 return new(CommonErrors.ItemAlreadyOwned(req.ItemId));
             }
 
+            if (user.Items.Any(ui => ui.Item!.Rank > 0))
```

Comment:
You're checking if the user owns an upgraded item when you want to check if the user is trying to buy an upgraded item.

Also you'll need to add a test for that case.

Reply:
Corrected and test created